04.lsコマンドを作る4の解答提出

yokomaru opened the pull request:
## 実装
### 必須要件
- [x] `-l`オプション付きのlsコマンドを作って提出する。

### 要件外の修正
- rubocop Metrics/PerceivedComplexity エラー修正のため`option -a`の実装をメソッドに切り出した 
https://github.com/yokomaru/ruby-practices/pull/7/commits/da765f3cae31519a00c208ba233acf5709b31119

yokomaru committed:
add: ls -lの結果と一致するかどうかのテストケースを追加

yokomaru committed:
add: option_paramsにlを追加

yokomaru committed:
add: ファイルのユーザーID、グループIDから表示名を出力するためEtcモジュールを追加

yokomaru committed:
add: ls-lで使用する定数とハッシュを追加

yokomaru committed:
fix: option l のブロックの中身がrになっていたため修正

yokomaru committed:
add: ロングフォーマットに関するメソッドを追加

yokomaru committed:
add: ファイル指定してlsコマンド実行時にロングフォーマットオプションの実装を反映

yokomaru committed:
add: ディレクトリ指定してlsコマンド実行時にロングフォーマットオプションの実装を反映

yokomaru committed:
fix: rubocop指摘箇所修正

yokomaru committed:
fix: rubocop指摘箇所解消のためlongformat_filesの作成、表示処理を関数に切り分ける

yokomaru committed:
fix: display_longformat_fileをgenerate_longformat_lineに変更＋マジックナンバーを定数に変更＋変数を修正

yokomaru committed:
fix: rubocop Metrics/PerceivedComplexity エラー修正ためdirectory_files作成処理を関数に切り出す

yokomaru committed:
fix: ロングフォーマットの表示関数をファイル指定時とディレクトリ指定時で分ける

yokomaru committed:
fix: generate_longformat_files(ファイルパスを呼び出し元で取得するように修正＋conversion_special_permissionの引数修正＋latest_modify_datetimeを月日時間でわける)

yokomaru committed:
fix: generate_longformat_file_line(メソッド名の修正、バッファ用のスペース数を定数に変更)

yokomaru committed:
add: ls用のtestディレクトリとファイル作成

yokomaru committed:
add: ls -l コマンド実行結果のテストを追加

yokomaru committed:
add: conversion_special_permissionのテスト追加

yokomaru committed:
fix: 関数名を修正(generate_longformat_file, generate_longest_bytesizes, convert_permission)

yokomaru committed:
fix: generate_longest_bytesizesで受け取った変数が仮の名称(hash)のままだったのでfilesに修正

yuuu commented on 04.ls/ls.rb:
```diff
@@ -94,8 +137,81 @@ def transpose_display_files(display_files)
+  longformat_files.each { |longformat_file| puts generate_longformat_file_line(longformat_file, longest_bytesizes) }
+end
+
+def generate_longformat_file(file, file_path)
```
Rubocopで特に指摘は出ていないとのことですが、このメソッドのコード量が多いように見えるので、処理の一部を別メソッドに切り出せないか検討をお願いします

yokomaru replied:
こちら以下のように処理を切り出しました。

- 修正前
```ruby
def generate_longformat_file(file, file_path)
  file_stat = File::Stat.new(file_path)
  file_stat_mode = file_stat.mode.to_s(8).rjust(6, '0')
  file_type = file_stat_mode.slice(0, 2)
  special_permission = file_stat_mode.slice(2)
  owner_permission_string = convert_permission(special_permission, file_stat_mode.slice(3), SUID_PERMISSION)
  group_permission_string = convert_permission(special_permission, file_stat_mode.slice(4), SGID_PERMISSION)
  other_permission_string = convert_permission(special_permission, file_stat_mode.slice(5), STICKEY_PERMISSION)
  {
    filemode: "#{FILE_TYPE[file_type]}#{owner_permission_string}#{group_permission_string}#{other_permission_string}",
    hardlink_nums: file_stat.nlink.to_s,
    owner_name: Etc.getpwuid(file_stat.uid).name,
    group_name: Etc.getgrgid(file_stat.gid).name,
    bytesize: file_stat.size.to_s,
    latest_modify_month: file_stat.mtime.strftime('%-m'),
    latest_modify_date: file_stat.mtime.strftime('%-d'),
    latest_modify_time: file_stat.mtime.strftime('%R'),
    filename: file,
    blocks: file_stat.blocks
  }
end
```

- 修正後
```rb
def generate_longformat_file(file, file_path)
  file_stat = File::Stat.new(file_path)
  {
    filemode: filemode(file_stat),
    hardlink_nums: file_stat.nlink.to_s,
    owner_name: Etc.getpwuid(file_stat.uid).name,
    group_name: Etc.getgrgid(file_stat.gid).name,
    bytesize: file_stat.size.to_s,
    latest_modify_datetime: file_stat.mtime.strftime('%_m %e %H:%M'),
    filename: file,
    blocks: file_stat.blocks
  }
end

def filemode(file_stat)
  file_stat_mode = file_stat.mode.to_s(8).rjust(6, '0')
  each_permissions = generate_each_permissions(file_stat_mode)
  each_permissions.map do |permissions|
    convert_permission(file_stat_mode[SPECIAL_PERMISSION_INDEX], permissions[0], permissions[1])
  end.unshift(FILE_TYPE[file_stat_mode[0, SPECIAL_PERMISSION_INDEX]]).join
end

def generate_each_permissions(file_stat_mode)
  [OWNER_PERMISSION_INDEX, GROUP_PERMISSION_INDEX, OTHER_PERMISSION_INDEX].map do |index|
    [file_stat_mode[index], index == OTHER_PERMISSION_INDEX ? STICKEY_PERMISSION : SUID_PERMISSION]
  end
end


```

yuuu commented on 04.ls/ls.rb:
```diff
@@ -94,8 +137,81 @@ def transpose_display_files(display_files)
+  file_stat_mode = file_stat.mode.to_s(8).rjust(6, '0')
+  file_type = file_stat_mode.slice(0, 2)
+  special_permission = file_stat_mode.slice(2)
+  owner_permission_string = convert_permission(special_permission, file_stat_mode.slice(3), SUID_PERMISSION)
```
ここから3行、似たようなコードが続いているのでイテレータでコードをまとめられないか検討してみてください。

yokomaru replied:
以下のように修正し処理をまとめました！

- イテレータでコードをまとめるために、convert_permissionに渡していた2つ目と3つ目の引数を配列にする処理を追加(generate_each_permissions)
- generate_each_permissionsでループしconvert_permissionを行う

```.rb
def filemode(file_stat)
  file_stat_mode = file_stat.mode.to_s(8).rjust(6, '0')
  each_permissions = generate_each_permissions(file_stat_mode)
  each_permissions.map do |permissions|
    convert_permission(file_stat_mode[SPECIAL_PERMISSION_INDEX], permissions[0], permissions[1])
  end.unshift(FILE_TYPE[file_stat_mode[0, SPECIAL_PERMISSION_INDEX]]).join
end

def generate_each_permissions(file_stat_mode)
  [OWNER_PERMISSION_INDEX, GROUP_PERMISSION_INDEX, OTHER_PERMISSION_INDEX].map do |index|
    [file_stat_mode[index], index == OTHER_PERMISSION_INDEX ? STICKEY_PERMISSION : SUID_PERMISSION]
  end
end
```

yuuu commented on 04.ls/ls.rb:
```diff
@@ -94,8 +137,81 @@ def transpose_display_files(display_files)
+
+def generate_longest_bytesizes(files)
+  {
+    hardlink_num: files.map { |file| file[:hardlink_nums].to_s.bytesize }.max,
```
ここもキー名が変わっているだけで残りはほぼ同じコードの繰り返し(  `.to_s` の有無が異なるが、全て `.to_s` を呼んでも大丈夫)なので、イテレータを使えないか検討ください。

yokomaru replied:
こちら以下のようにキー名でmap処理を行い最大文字数のハッシュを作成するように修正いたしました。

```.rb
def generate_longest_bytesizes(files)
  # hardlink_nums owner_name group_name bytesizeはファイル名の最大文字数で右詰めするため取得
  %i[hardlink_nums owner_name group_name bytesize].each_with_object({}) do |sym, hash|
    hash[sym] = files.map { |file| file[sym].to_s.bytesize }.max
  end
end
```

yuuu commented on 04.ls/ls.rb:
```diff
@@ -94,8 +137,81 @@ def transpose_display_files(display_files)
+end
+
+def generate_longformat_file_line(longformat_file, longest_bytesizes)
+  "#{longformat_file[:filemode]} "\
```
ここは全て文字列とするよりは

```ruby
[
  longformat_file[:filemode],
  longformat_file[:hardlink_nums].rjust(longest_bytesizes[:hardlink_num] + BUFFER_WIDTH),
  # 省略
].join(' ')
```

のようにした方がコードの量を減らせそうですね。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -94,8 +137,81 @@ def transpose_display_files(display_files)
+  return PERMISSION[permission] if special_permission != target_permission
+
+  if permission.to_i.odd?
+    PERMISSION[permission].sub(/x$/,
```
ここは `sub` の引数のみを切り替えるようにすると、コードの重複を減らせますね。

yokomaru replied:
PERMISSION[permission]の最後の文字を置き換えたいだけだったので、
subからchopメソッドに変更しpermissionの数値次第で大文字小文字を切り替えて結合するように修正いたしました！

```.rb
def convert_permission(special_permission, permission, target_permission)
  return PERMISSION[permission] if special_permission != target_permission

  special_permission_character = permission.to_i.odd? ? SPECIAL_PERMISSION[special_permission] : SPECIAL_PERMISSION[special_permission].upcase
  PERMISSION[permission].chop + special_permission_character
end
```

yuuu commented on 04.ls/ls.rb:
```diff
@@ -94,8 +137,81 @@ def transpose_display_files(display_files)
+  }
+end
+
+def generate_longest_bytesizes(files)
```
メソッド名のプレフィックス `generate_` は事実上 `get_` と同じ使われ方をしていて、Rubyだとこのプレフィックスは省略することが多いですね。省略しても意味は通じると思いました。

yokomaru replied:
ありがとうございます、1点悩んでいるので質問させてください！
上記コメントいただいた箇所のプレフィックスを省略すると、以下のように変数名とメソッド名が同じになります。

```.rb
# 修正前
longest_bytesizes = generate_longest_bytesizes(longformat_files)

# 修正後
longest_bytesizes = longest_bytesizes(longformat_files)
```

以前カレンダー課題で[メソッド名と同じ名前のローカル変数を宣言しないほうが良い](https://qiita.com/jnchito/items/fb438aa0786bb3cbf919)という指摘をいただき、そこからなるべく値をかえすメソッドにはgenerate などをつけて命名するようにしております。

ただ、ご指摘と以下の記事にもあるようにこのメソッドは単純に値を返すメソッドなので省略しても名詞で問題ないと思っております。

[注意：動詞で始めるのは、あくまで「処理を実行するメソッド」が対象です
](https://qiita.com/jnchito/items/459d58ba652bf4763820#%E6%B3%A8%E6%84%8F%E5%8B%95%E8%A9%9E%E3%81%A7%E5%A7%8B%E3%82%81%E3%82%8B%E3%81%AE%E3%81%AF%E3%81%82%E3%81%8F%E3%81%BE%E3%81%A7%E5%87%A6%E7%90%86%E3%82%92%E5%AE%9F%E8%A1%8C%E3%81%99%E3%82%8B%E3%83%A1%E3%82%BD%E3%83%83%E3%83%89%E3%81%8C%E5%AF%BE%E8%B1%A1%E3%81%A7%E3%81%99)

上記を踏まえた上で、適した変数名と関数名を考えているのですがいまいちいいものが浮かばず、
yuuuさんだったらどのように命名するかアドバイスをいただきたいです・・！

yuuu replied:
なるほど、ぜひこのDocsを読んでいただきたいです 🙏 

https://bootcamp.fjord.jp/pages/avoid-too-specific-var-name

このDocsに照らし合わせて考えると、変数名の `longest_bytesizes` は `bytesizes` でも特に差し支えがないと思います 👍

yokomaru replied:
コメントありがとうございます！
Docsとても参考になりました！関連記事のyuuuさんの以下記事もとても勉強になりましたーー！
[名前重要！具体的過ぎるメソッド名や変数名の弊害と適切な抽象度 \| FJORD BOOT CAMP（フィヨルドブートキャンプ）](https://bootcamp.fjord.jp/articles/78)

命名に関して悩んでいた理由として、この処理がどういう用途で使われるべきなのかをちゃんと言語化できていなかったからだと思ったので、改めてどういう用途で使っているかを考え直し今回追加した各メソッド名を修正いたしました。

#### `generate_longest_bytesizes`
- 各要素の最大のバイト数の長さのハッシュを作成する処理
##### 変更後
- メソッド名
   `longest_bytesizes`
- 変数名
   `bytesizes `

#### `generate_each_permissions`
- convert_permissionの引数として使用する各パーミッションの８進数の数値たちのハッシュを作成する処理

##### 変更後
- メソッド名
`each_octal_permissions`

- 変数名
`octal_permissions`

#### `generate_longformat_file`
- 各ファイルのlsコマンドの表示に必要なステータス情報を取得する処理

##### 変更後
- メソッド
`file_statuses`
- 変数名
` statuses` 

#### `generate_longformat_file_line`
- lsコマンドで実際に表示するロングフォーマットの情報を作成する処理

##### 変更後
`longformat_files`

yokomaru committed:
fix: generate_longformat_fileを分割

filemodeに必要な情報を取得・作成している処理をfilemodeメソッドに分ける

yokomaru committed:
fix: generate_longformat_file_lineメソッドの修正

文字列結合していたところを配列にしてjoinメソッドで結合

yokomaru committed:
fix: generate_longest_bytesizes

同じような処理の重複をシンボルをキーでループさせて処理を行う

yokomaru committed:
fix: convert_permission

subで最後の文字を変換かけていたところをchop+特殊文字の結合処理に変更

yokomaru committed:
add: generate_longest_bytesizesにコメント追加

イテレーターに処理するように変更したので、最大文字数で文字つめが必要なキーで処理していることを記載

yokomaru committed:
update: latest_modify_month、latest_modify_timeを削除しlatest_modify_dateに統一

strftime('%_m %e %H:%M')で書き換えられることがわかったため修正

yokomaru committed:
update: generate_longformat_file_lineをイテレーターで書き換え

不要になった定数を削除

yokomaru committed:
update: generate_each_permissionsを追加

filemodeからeach_permissionsの配列を作成する処理を切り出す

yokomaru committed:
update: generate_each_permissions

配列をイテレーターで作成しコードの重複を削除
パーミッションの各位置を定数で持たせる

yokomaru committed:
update: generate_longformat_file_line

map内の変数を修正、コメントの追加

yokomaru committed:
update: generate_longformat_file_line

rubocop Layout/LeadingCommentSpace: Missing space after #. の修正

yokomaru committed:
fix: generate_longformat_file_line

join後に末尾に空白ができてしまっていたのでstripで取り除いて返却する

yokomaru committed:
fix: method and variable names

不要なgenerateプレフィックスがついたメソッド名の修正
他の変数と区別の必要ない変数名を修正

yokomaru committed:
fix: variable names in total block count process

statuses配列内のblocksの合計処理を計算するブロック内変数がfileになっていたのでstatusに変更

yokomaru committed:
update: display_longformat_files

display_longformat_filesとdisplay_longformat_directory_filesを統一
処理の重複を削除するためdisplay_longformat_files内でmapやeachしていた処理ごと切り出してメソッド化する

yokomaru committed:
update: method name change permissions_and_target_permissions to permissions_and_target_permissions

filemodeの文字列から８進数の数値に切り出すためメソッド名をeach_octal_permissionsに修正
convert_permissionにわたす引数をpermissionsというハッシュにまとめて渡すように修正＋テストも併せて修正

yokomaru committed:
fix: ls_test.rb rubocop

Layout/ExtraSpacing: Unnecessary spacing detected
Layout/SpaceInsideHashLiteralBraces

yuuu commented on 04.ls/ls.rb:
```diff
@@ -32,29 +69,37 @@ def main
 end
 
-def display_files(files)
+def display_files(files, options)
```
このメソッドには `options` を渡す仕様ですが、今のところ `-l` しか使わないのでそれだけを渡す仕様としたほうがYAGNI原則に沿っていて良いですね。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -94,8 +139,82 @@ def transpose_display_files(display_files)
+def convert_permission(permissions)
+  return PERMISSION_TYPE[permissions[:permission]] if permissions[:special_permission] != permissions[:target_permission]
+
+  character = if permissions[:permission].to_i.odd?
```
ここは `odd?` じゃなければ `.upcase` する、という書き方にすると少しだけコードをシンプルにできそうですね。

yokomaru replied:
>odd? じゃなければ .upcase する

even?がtrureなら.upcaseするとして以下のように書き換えました。

```
  special_permission_type = SPECIAL_PERMISSION_TYPE[special_permission]
  special_permission_type = special_permission_type.upcase if permission.to_i.even?
```

yuuu commented on 04.ls/ls.rb:
```diff
@@ -94,8 +139,82 @@ def transpose_display_files(display_files)
+end
+
+def convert_permission(permissions)
+  return PERMISSION_TYPE[permissions[:permission]] if permissions[:special_permission] != permissions[:target_permission]
```
`permissions[:special_permission]` が複数回登場するので、最初に変数に入れるようにすると少しだけコードをシンプルにできそうです。

yokomaru replied:
permissionsのハッシュを作成する処理を削除したため、引数を3つ渡すように修正しました。
複数登場していた` PERMISSION_TYPE[permission]`などの定数から値を持ってくる箇所を変数に入れました。

```rb
def filemode(status)
  mode = status.mode.to_s(8).rjust(6, '0')
  [OWNER_PERMISSION_INDEX, GROUP_PERMISSION_INDEX, OTHER_PERMISSION_INDEX].map do |index|
    convert_permission(mode[index], mode[SPECIAL_PERMISSION_INDEX], TARGET_SPECIAL_PERMISSION[index])
  end.unshift(FILE_TYPE[mode[0, SPECIAL_PERMISSION_INDEX]]).join
end
```

```rb
def convert_permission(permission, special_permission, target_special_permission)
  permission_type = PERMISSION_TYPE[permission]
  return permission_type if special_permission != target_special_permission

  special_permission_type = SPECIAL_PERMISSION_TYPE[special_permission]
  special_permission_type = special_permission_type.upcase if permission.to_i.even?
  [permission_type.chop, special_permission_type].join
end
```

yuuu commented on 04.ls/ls.rb:
```diff
@@ -94,8 +139,82 @@ def transpose_display_files(display_files)
+    status.reject { |key| key == :blocks }.map do |key, value| # blocksは表示には使用しないため表示の配列から除く
+      # filemode owner_name group_name は右隣と２スペース分空いているため空白文字を追加
+      buffer_space = ' ' if %i[filemode owner_name group_name].include?(key)
+      bytesizes[key] ? value.rjust(bytesizes[key]) + buffer_space.to_s : value + buffer_space.to_s
```
`+ buffer_space.to_s` は三項演算子の外に出せますね。
`bytesizes[key].to_i` すれば三項演算子自体なくせそうにも見えます 👀

yokomaru replied:
おっしゃる通り三項演算子が不要になりました！

- bytesizes[key].to_iでnilの場合は0になる
- rjust(0)など文字列よりrjust内の数字が小さい場合はselfの文字列をそのまま返す

また、 buffer_space.to_s はbuffer_spaceがnilだった時のためにto_sしていたのですが、
以下のように配列にしてjoinさせたら不要になりましたので削除いたしました。

`[value.rjust(bytesizes[key].to_i),  buffer_space].join`

yokomaru committed:
fix: only option l display_files method argument

メソッド内で Lオプションしか使用しないので引数を限定して渡す

yokomaru committed:
fix: set variables permissions

コードを簡略化するため、複数使用するpermissionsハッシュを変数に入れる

yokomaru committed:
fix: convert_permission

filemodeメソッドから渡す引数を最初から分けて渡す
ハッシュ定数から値を取得する処理が重複するので変数に入れる
odd?で分岐していたところをeven?に変更し分岐を少なくする
文字列結合を+から.joinに変更

yokomaru committed:
fix: longformat_files

bytesizes[key]での分岐を削除
文字列結合をjoinに変更
stripが不要のため削除

yokomaru committed:
fix: convert_permission test

引数の数を変更

yokomaru committed:
add: TARGET_SPECIAL_PERMISSION constant

each_octal_permissionsのtarget_permissionの取得部分での分岐をハッシュに修正

yokomaru committed:
fix: filemode and delete each_octal_permissions

each_octal_permissionsの結果でループすることで２重でループが走っていたのでfilemode内で1回のループで処理するように修正
convert_permission内での引数の受け渡しの順番を修正
特殊権限のチェックに使っているtarget_permission引数が意味が分かりづらかったのでtarget_special_permissionに変更
テストの修正

yokomaru commented:
こちらOKいただいたのでマージいたします！
